compiler/rustc_resolve/src/macros.rs

```diff
@@ -341,12 +341,16 @@ impl<'ra, 'tcx> ResolverExpand for Resolver<'ra, 'tcx> {
     fn record_macro_rule_usage(&mut self, id: NodeId, rule_i: usize) {
         let did = self.local_def_id(id);
         if let Some(rules) = self.unused_macro_rules.get_mut(&did) {
-            rules.remove(&rule_i);
+            rules.shift_remove(&rule_i);
```

Comment:
I didn't notice it before, but this is also expensive, similarly to `unused_(labels,macros)`, because all rules are added and then removed on use.
`UnordMap` and sorting before reporting is recommended.

Reply:
Same as [this](https://github.com/rust-lang/rust/pull/131213/files#r1817995304). Am I supposed to implement `StableCompare` for the key? I'm not sure because this seems like a lot of work compared to using IndexMap or converting directly to vector.

Reply:
Addressed here: https://github.com/rust-lang/rust/pull/131213/commits/5d14c349afd52809b0dc35815cc53992c0b3fef3
Please check.